router/pkg/connectrpc/handler.go

```diff
@@ -0,0 +1,629 @@
+		zap.String("service", serviceName),
+		zap.String("method", methodName))
+
+	err := h.validator.ValidateMessage(serviceName, methodName, requestJSON)
+	if err == nil {
+		h.logger.Debug("proto validation passed",
+			zap.String("service", serviceName),
+			zap.String("method", methodName),
+		)
+		return nil
+	}
+
+	var validationErr *ValidationError
+	if errors.As(err, &validationErr) {
+		h.logger.Warn("proto validation failed",
+			zap.String("service", serviceName),
+			zap.String("method", methodName),
+			zap.String("error", validationErr.Error()))
+		return connect.NewError(
+			connect.CodeInvalidArgument,
+			fmt.Errorf("invalid request: %s", validationErr.Error()),
+		)
+	}
+
+	// For other errors, log and continue (don't block on validation errors)
+	h.logger.Warn("validation check failed, continuing",
+		zap.String("service", serviceName),
+		zap.String("method", methodName),
+		zap.Error(err))
+	return nil
+}
+
+// HandleRPC processes an RPC request and returns a response
+// serviceName: fully qualified service name (e.g., "mypackage.MyService")
+// methodName: the RPC method name (e.g., "GetUser" or "QueryGetUser")
+// requestJSON: the JSON-encoded request body
+// ctx: request context with headers
+func (h *RPCHandler) HandleRPC(ctx context.Context, serviceName, methodName string, requestJSON []byte) ([]byte, error) {
+	h.logger.Debug("handling RPC request",
+		zap.String("service", serviceName),
+		zap.String("method", methodName),
+		zap.String("request_json", string(requestJSON)))
+
+	// Validate the proto message structure before processing
+	if err := h.validateRequest(serviceName, methodName, requestJSON); err != nil {
+		return nil, err
+	}
+
+	// Look up operation from registry scoped to this service
+	// This ensures operations can only be called from their owning service
+	// The method name must exactly match the operation name
+	operation := h.operationRegistry.GetOperationForService(serviceName, methodName)
+	if operation == nil {
+		// Log all available operations for this service to help diagnose the issue
+		allOps := h.operationRegistry.GetAllOperationsForService(serviceName)
+		var availableOps []string
+		for _, op := range allOps {
+			availableOps = append(availableOps, op.Name)
+		}
+		h.logger.Error("operation not found",
+			zap.String("service", serviceName),
+			zap.String("requested_method", methodName),
+			zap.Strings("available_operations", availableOps))
+		return nil, fmt.Errorf("operation not found for service %s: %s", serviceName, methodName)
+	}
+
+	h.logger.Debug("resolved operation",
+		zap.String("service", serviceName),
+		zap.String("rpc_method", methodName),
+		zap.String("operation", operation.Name),
+		zap.String("type", operation.OperationType))
+
+	// Convert proto JSON (snake_case) to GraphQL variables (camelCase)
+	// Pass service and method info for schema-aware enum detection
+	variables, err := h.convertProtoJSONToGraphQLVariables(serviceName, methodName, requestJSON)
```

Comment:
I don't understand why this is happening. Are you assuming that proto is always snake and GraphQL is always camelCase? I don't think we can make such assumptions.